src/coreclr/gc/env/gcenv.os.h

```diff
@@ -410,6 +410,11 @@ class GCToOSInterface
     //  non zero if it has succeeded, 0 if it has failed
     static size_t GetVirtualMemoryLimit();
 
+    // Return the maximum address of the of the virtual address space of this process.
+    // Return:
+    //  non zero if it has succeeded, 0 if it has failed
+    static size_t GetVirtualMemoryMaxAddress();
```

Comment:
we already needed to consider the virtual address space for 32-bit. can we modify that to accommodate this instead of adding a new interface method? please see [this](https://github.com/dotnet/runtime/blob/main/src/coreclr/gc/windows/gcenv.windows.cpp#L1039)

we could just make it so that the limit is returned in `GetPhysicalMemoryLimit`. do you see a problem with doing that?

Reply:
@Maoni0 I am not sure if setting the physical memory limit to the virtual memory limit is sufficient. I have to admit though that there are various calculations using the limits and I am not 100% sure I have a clear picture of all of them.
However, I have tried to apply your suggestion and found that it doesn't work for regions. The following code sets the regions range to 5 times the heap_hard_limit
```
   45270                        // We use this calculation because it's close to what we used for segments.
-> 45271                        gc_heap::regions_range = ((gc_heap::use_large_pages_p) ? (2 * gc_heap::heap_hard_limit)
   45272                                                                               : (5 * gc_heap::heap_hard_limit));
```

which obviously results in OOM due to the virtual memory limit getting exceeded here:
```
  13684                size_t reserve_size = regions_range;
-> 13685                uint8_t* reserve_range = (uint8_t*)virtual_alloc (reserve_size, use_large_pages_p);
   13686                if (!reserve_range)
   13687                    return E_OUTOFMEMORY;
```
I this that's the reason why @cshung has suggested to make the change to limit the `gc_heap::regions_range` by the amount of available virtual memory space.

Reply:
@cshung is that the reason? if we can't have 5x heap_hard_limit due to the VM space, we should just not try to reserve 5x the hard_limit (just like we do with large pages).

regardless, we need to care about VM in the implementation of `GetMemoryStatus` because if VM is the limiting factor we'd want to know that and perform more aggressive GCs if necessary in that case. I don't believe that current change takes care of that, correct?

Reply:
@Maoni0, Yes, in the case when virtual memory is limited, we need to adjust our region's range. The region's range has to be less than how much virtual memory we have. I don't think we have an option here.

I didn't mean that change is complete for addressing the virtual-memory limited scenario, it is just to get started.

Now I think about it, perhaps we should have an implied `heap_hard_limit` based on the virtual memory limit and let our existing `heap_hard_limit` based heuristic to do the work for us.